test/units/testsuite-22.12.sh

```diff
@@ -0,0 +1,154 @@
+#! /bin/bash
+
+# Test the "Age" parameter (with age-by) for systemd-tmpfiles.
+
+set -e
+set -x
+
+# Test directory structure looks like this:
+#   /tmp/ageby/
+#   ├── d1
+#   │   ├── f1
+#   │   ├── f2
+#   │   ├── f3
+#   │   └── f4
+#   ├── d2
+#   │   ├── f1
+#   │   ├── f2
+#   ...
+
+export SYSTEMD_LOG_LEVEL="debug"
+
+rm -rf /tmp/ageby/d{1..5}
+mkdir -p /tmp/ageby/d{1..5}
+
+touch -a --date "2 minutes ago" /tmp/ageby/d1/f1
+touch -m --date "4 minutes ago" /tmp/ageby/d2/f1
+
+# Create a bunch of other files.
+touch /tmp/ageby/d{1,2,5}/f{2..4}
+
+# For "ctime".
+touch /tmp/ageby/d3/f1
+chmod +x /tmp/ageby/d3/f1
+sleep 3
+
+# For "btime".
+touch /tmp/ageby/d4/f1
+sleep 3
+
+# More files with recent "{a,b}time" values.
+touch /tmp/ageby/d{3,4}/f{2..4}
+
+# Check for cleanup of "f1" in each of "/tmp/d{1..4}".
+time systemd-tmpfiles --clean - <<-EOF
+d /tmp/ageby/d1 - - - a:1m -
+e /tmp/ageby/d2 - - - m:3m -
+D /tmp/ageby/d3 - - - c:4s -
+d /tmp/ageby/d4 - - - b:2s -
```

Comment:
(Please ignore the `time` command there.)

I'm at a loss here. This is from the logs from the CI build:
```
Running clean action for entry d /tmp/ageby/d4
Cleanup threshold for directory "/tmp/ageby/d4" is Fri 2021-05-07 14:29:17.103380 UTC (age-by: btime (f))
File "/tmp/ageby/d4/f4": modify time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f4": access time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f4": change time Fri 2021-05-07 14:29:19.083394 UTC is too new.
Removing "/tmp/ageby/d4/f4".
File "/tmp/ageby/d4/f3": modify time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f3": access time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f3": change time Fri 2021-05-07 14:29:19.083394 UTC is too new.
Removing "/tmp/ageby/d4/f3".
File "/tmp/ageby/d4/f2": modify time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f2": access time Fri 2021-05-07 14:29:19.083394 UTC is too new.
File "/tmp/ageby/d4/f2": change time Fri 2021-05-07 14:29:19.083394 UTC is too new.
Removing "/tmp/ageby/d4/f2".
Removing "/tmp/ageby/d4/f1".
Restoring access and modification time on "/tmp/ageby/d4": Fri 2021-05-07 14:29:13.073608 UTC, Fri 2021-05-07 14:29:19.083394 UTC
```

And this is when I test is locally:
```
Running clean action for entry d /tmp/ageby/d4
Cleanup threshold for mount point "/tmp/ageby/d4" is Fri 2021-05-07 16:40:55.471216 UTC (age-by: btime (f))
File "/tmp/ageby/d4/f4": modify time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f4": access time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f4": change time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f4": birth time Fri 2021-05-07 16:40:57.444339 UTC is too new.
Removing "/tmp/ageby/d4/f1".
File "/tmp/ageby/d4/f2": modify time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f2": access time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f2": change time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f2": birth time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f3": modify time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f3": access time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f3": change time Fri 2021-05-07 16:40:57.444339 UTC is too new.
File "/tmp/ageby/d4/f3": birth time Fri 2021-05-07 16:40:57.444339 UTC is too new.
Restoring access and modification time on "/tmp/ageby/d4": Fri 2021-05-07 16:40:51.420339 UTC, Fri 2021-05-07 16:40:57.444339 UTC
```

I may have a guess about why it fails in CI, but passes locally. Something seems to make it think `btime` of all the files in `/tmp/ageby/d4` is older than the cut-off time.

Is it perhaps ...
```
btime_nsec = FLAGS_SET(sx.stx_mask, STATX_BTIME) ? load_statx_timestamp_nsec(&sx.stx_btime) : 0;
```
... and `FLAGS_SET(sx.stx_mask, STATX_BTIME)` is evaluating to a false?

Should I check for that when `b` is specified for the `age-by` argument (or for all `statx` timestamp fields)?

@poettering  - Can you please advise? Thanks.

Reply:
So, I made a few changes to the tests (to disable `btime` checks) when it is not available. Tests are passing.